crates/hotshot-state-prover/src/circuit.rs

```diff
@@ -0,0 +1,123 @@
+//! Circuit implementation for verifying light client state update
+
+use std::marker::PhantomData;
+
+use ark_ec::twisted_edwards::TECurveConfig;
+use ethereum_types::U256;
+use hotshot_types::traits::stake_table::{SnapshotVersion, StakeTableScheme};
+use jf_plonk::errors::PlonkError;
+use jf_primitives::{
+    circuit::signature::schnorr::VerKeyVar,
+    rescue::RescueParameter,
+    signatures::{
+        bls_over_bn254::VerKey as BLSVerKey,
+        schnorr::{Signature, VerKey as SchnorrVerKey},
+    },
+};
+use jf_relation::gadgets::ecc::TEPoint;
+use jf_relation::{
+    errors::CircuitError, gadgets::ecc::SWToTEConParam, BoolVar, Circuit, PlonkCircuit,
+};
+
+use self::{
+    config::{u256_to_field, NUM_ENTRIES},
+    types::{HotShotState, StakeTableEntryVar},
+};
+
+pub mod config;
+pub mod types;
+
+#[derive(Clone, Debug)]
+pub struct StateUpdateBuilder<F: RescueParameter>(PhantomData<F>);
+
+impl<F> StateUpdateBuilder<F>
+where
+    F: RescueParameter + SWToTEConParam,
+{
+    /// A function that takes as input:
+    /// - stake table entries (`Vec<(BLSVerKey, SchnorrVerKey, Amount)>`)
+    /// - schnorr signatures of the updated states (`Vec<SchnorrSignature>`)
+    /// - updated hotshot state (`(view_number, block_height, block_comm, fee_ledger_comm, stake_table_comm)`)
+    /// - signer bit vector
+    /// - quorum threshold
+    /// checks that
+    /// - the signer's accumulated weight exceeds the quorum threshold
+    /// - the commitment of the stake table
+    /// - all schnorr signatures are valid
+    pub fn build<ST, P>(
+        stake_table: &ST,
+        _sigs: &[Signature<P>],
+        _hotshot_state: &HotShotState<F>,
+        signer_bit_vec: &[bool],
+        threshold: &U256,
+    ) -> Result<PlonkCircuit<F>, PlonkError>
+    where
+        ST: StakeTableScheme<Key = (BLSVerKey, SchnorrVerKey<P>), Amount = U256>,
+        P: TECurveConfig<BaseField = F>,
+    {
+        let mut circuit = PlonkCircuit::new_turbo_plonk();
```

Comment:
What are the public input of the circuit? I thought it's only
1. The stake table commitment
2. The hotshot state.
3. The threshold.

All others could be secret witness.

Reply:
Done in the newest commit.